Add ability to subscribe to multiple keys and to prevent propagation

fredyshox opened the pull request:
Attempt to fix #30

Added `CompositeUserDefaultsKeyObservation` for observing multiple keys, with `preventPropagation` option to prevent infinite recursion when change is made within handler. 

I store flag in threadDictionary of the current thread to determine when changes should be propagated. This enables to still support changes made on other threads, while initial one is still processing handler (more in `testObservePreventPropagationMultiThread` test case).

Code is still a little messy, but let me know what you think :)


<!-- Issuehunt content -->

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#30: Add ability to subscribe to multiple keys or any change in a given suite](https://issuehunt.io/repos/129132562/issues/30)
---
</details>
<!-- /Issuehunt content-->

fredyshox commented:
As for `observeAll` and `publisherAll` for all known keys, `Mirror` does not yet support static fields, so my initial thought of `Mirror(Defaults.Keys.self).children` doesn't work.

sindresorhus commented on Sources/Defaults/Observation.swift:
```diff
@@ -268,6 +363,23 @@ extension Defaults {
 	}
+	
+	public static func observe(
+		keys: DefaultsBaseKey...,
```
I think https://github.com/sindresorhus/Defaults/issues/50 applies here too...

fredyshox replied:
If we added `DefaultsBaseKey` conformance to `Keys` and use it here, dot syntax would work. That would make `Keys` an abstract class and `DefaultsBaseKey` redundant.

sindresorhus replied:
Alright. Let's do that.

sindresorhus commented on Sources/Defaults/Defaults.swift:
```diff
@@ -1,12 +1,12 @@
 import Foundation
 
-public protocol _DefaultsBaseKey: Defaults.Keys {
+public protocol DefaultsBaseKey: Defaults.Keys {
```
Can you add `typealias BaseKey = DefaultsBaseKey` to line 17 and then use `BaseKey` throughout?

sindresorhus commented:
This is looking promising. Thanks for working on it. 🙌

sindresorhus commented:
> As for observeAll and publisherAll for all known keys, Mirror does not yet support static fields, so my initial thought of Mirror(Defaults.Keys.self).children doesn't work.

Too bad. "static" always lags behind in functionality. We could actually solve this without mirror if we switch to using Key paths instead of the dot syntax (It would then be `Defaults[\.foo]`. But of course "static key paths" are not yet implemented in Swift.

sindresorhus commented:
Is there any way we could add propagation prevention to `publisher(keys:)` too?

fredyshox commented:
> Is there any way we could add propagation prevention to `publisher(keys:)` too?

We could do that, by tweaking `DefaultsSubscription` into using new `CompositeUserDefaultsKeyObservation`. But there're few catches. Note that `preventPropagation` won't protect you, if you use `dispatch.async` inside observation handler. And in Combine is far more common to use things like `receive(on:)`, `debounce`, `delay` etc.

fredyshox commented:
I’ve came up with a different approach, to make this feasible on both classic and combine implementations. A block based approach inspired by SwiftUIs `withAnimation`:

```swift
func withNoPropagation(block: () -> Void) {
	// tweak threadDictionary and call block
}
```

Every change made in the block, would not be propagated at all. Usage example:

```swift
Defaults.observeAll(.key1, .key2) {
	// …
	withNoPropagation {
		// update some value at .key1
		// this will not be propagated
		Defaults[.key1] = 11
	}
	// this will be propagated
	Defaults[.someKey] = true 
}
``` 

A big advantage is that, this may be used not only in observation callback, but also in Combine callbacks (regardless of publisher modifiers) or any other place. 

I’ve working implementation, let me know what you think 👍.

sindresorhus commented:
@fredyshox That's a good approach! I like it.

Should it be `withNoPropagation` or `withoutPropagation`?

sindresorhus commented on Sources/Defaults/Observation.swift:
```diff
@@ -140,6 +140,17 @@ extension Defaults {
 		}
 	}
+	
+	static var preventPropagationThreadDictKey: String {
```
This should be `private`.

sindresorhus commented on Sources/Defaults/Observation.swift:
```diff
@@ -140,6 +140,17 @@ extension Defaults {
 		}
 	}
+	
+	static var preventPropagationThreadDictKey: String {
+		return "\(type(of: Observation.self))_threadUpdatingValuesFlag"
+	}
```
```suggestion
	static var preventPropagationThreadDictKey: String {
		"\(type(of: Observation.self))_threadUpdatingValuesFlag"
	}
```

sindresorhus commented on Sources/Defaults/Observation.swift:
```diff
@@ -140,6 +140,17 @@ extension Defaults {
+	}
+	
+	public static func withoutPropagation(block: () -> Void) {
+		let key = Defaults.preventPropagationThreadDictKey
```
```suggestion
		let key = preventPropagationThreadDictKey
```

Same in other places.

sindresorhus commented on Sources/Defaults/Observation.swift:
```diff
@@ -200,11 +211,108 @@ extension Defaults {
+			
+			
+			let key = preventPropagationThreadDictKey
+			let updatingValuesFlag = (Thread.current.threadDictionary[key] as? Bool) ?? false
```
```suggestion
			let updatingValuesFlag = Thread.current.threadDictionary[key, default: false]
```

fredyshox replied:
This is `NSMutableDictioary`, so it doesn't support that.

sindresorhus commented on Sources/Defaults/Observation.swift:
```diff
@@ -200,11 +211,108 @@ extension Defaults {
+			
+			let key = preventPropagationThreadDictKey
+			let updatingValuesFlag = (Thread.current.threadDictionary[key] as? Bool) ?? false
+			if updatingValuesFlag {
```
```suggestion
			guard !updatingValuesFlag else {
```

sindresorhus commented on Sources/Defaults/Observation.swift:
```diff
@@ -200,11 +211,108 @@ extension Defaults {
+	final class CompositeUserDefaultsKeyObservation: NSObject, Observation {
+		private static var observationContext = 0
+		
+		final class SuiteKeyPair {
```
```suggestion
		private final class SuiteKeyPair {
```

sindresorhus commented on Sources/Defaults/Observation.swift:
```diff
@@ -200,11 +211,108 @@ extension Defaults {
 		}
 	}
+	
+	final class CompositeUserDefaultsKeyObservation: NSObject, Observation {
```
```suggestion
	private final class CompositeUserDefaultsKeyObservation: NSObject, Observation {
```

sindresorhus commented on Sources/Defaults/Observation.swift:
```diff
@@ -200,11 +211,108 @@ extension Defaults {
+			
+			let key = preventPropagationThreadDictKey
+			let updatingValuesFlag = (Thread.current.threadDictionary[key] as? Bool) ?? false
+			if updatingValuesFlag {
```
Same as the previous comment.

sindresorhus commented on Sources/Defaults/Observation.swift:
```diff
@@ -268,6 +376,22 @@ extension Defaults {
 		return observation
 	}
+	
+	public static func observe(
```
This needs a doc comment

sindresorhus commented:
Also replied to https://github.com/sindresorhus/Defaults/pull/49#discussion_r467627037.

sindresorhus commented:
Sorry for the late reply. Summer and COVID made it hard to keep up with all the PRs.

sindresorhus commented:
Can you document `withoutPropagation`?

fredyshox commented:
I've added docstrings to new public API and updated readme. Also added short comment explaining inner workings of `preventPropagation` inside its implementation 🙂.

sindresorhus commented:
This is looking great. Nice work as always 🙌🏻

sindresorhus commented:
@fredyshox Would you be interested in joining the project as a maintainer? There's no commitment needed. Just want to highlight your contributions with your name in the readme. No worries if not though.

fredyshox commented:
Yes I do! Feel free to mention me :)
